lib/api/objectDelete.js

```diff
@@ -23,44 +29,132 @@ export default function objectDelete(authInfo, request, log, cb) {
+        });
+        return cb(decryptVidResult);
+    }
+    const reqVersionId = decryptVidResult;
+
     const valParams = {
         authInfo,
         bucketName,
         objectKey,
+        versionId: reqVersionId,
         requestType: 'objectDelete',
-        log,
     };
-    return services.metadataValidateAuthorization(valParams,
-        (err, bucket, objMD) => {
-            const corsHeaders = collectCorsHeaders(request.headers.origin,
-                request.method, bucket);
-            if (err) {
-                log.debug('error processing request', {
-                    error: err,
-                    method: 'metadataValidateAuthorization',
+
+    const canonicalID = authInfo.getCanonicalID();
+    return async.waterfall([
+        function validateBucketAndObj(next) {
+            return metadataValidateBucketAndObj(valParams, log,
+            (err, bucketMD, objMD) => {
+                if (err) {
+                    return next(err, bucketMD);
+                }
+                const versioningCfg = bucketMD.getVersioningConfiguration();
+                if (!objMD) {
+                    if (!versioningCfg) {
+                        return next(errors.NoSuchKey, bucketMD);
+                    }
+                    // TODO: currently we return do not create a delete
+                    // marker if the version specified for deletion does
+                    // not exist. To adhere to AWS behavior, we should
+                    // update to create a delete marker anyway.
+                    if (reqVersionId) {
+                        const corsHeaders =
+                            collectCorsHeaders(request.headers.origin,
+                                request.method, bucketMD);
+                        return cb(null, corsHeaders);
+                    }
+                    // To adhere to AWS behavior, create a delete marker even
+                    // if trying to delete an object that does not exist when
+                    // versioning has been configured
+                    return next(null, bucketMD, objMD);
+                }
+                const headerValResult = validateHeaders(objMD,
+                    request.headers);
+                if (headerValResult.error) {
+                    return next(headerValResult.error, bucketMD);
+                }
+                if (objMD['content-length']) {
+                    log.end().addDefaultFields({
+                        contentLength: objMD['content-length'],
+                    });
+                }
+                return next(null, bucketMD, objMD);
+            });
+        },
+        function getVersioningInfo(bucketMD, objectMD, next) {
+            return preprocessingVersioningDelete(bucketName,
+                bucketMD, objectKey, objectMD, reqVersionId, log,
+                (err, options) => {
+                    if (err) {
+                        log.debug('err processing versioning info',
```

Comment:
This should be error level because at this point we don't expect the request to fail due to user input.